pkg/hasher/mutex.go

```diff
@@ -0,0 +1,24 @@
+package hasher
+
+import (
+	"sync"
+)
+
+// MutexHasher wraps a Hasher with a sync.Mutex to ensure thread-safe access.
```

Comment:
Are `Hasher` objects not thread-safe? That seems unexpected.

Is there a situation where we don't want them to be thread-safe?

Reply:
Correct, Hasher objects are generally not thread-safe.

There are situations where we don't want Hasher objects to be thread-safe, particularly In our case, each worker would have its own instance of a Hasher, ensuring no race conditions, and allowing us to bypass mutexes entirely. By doing this, we avoid the overhead of locking mechanisms, resulting in a ~20x speedup. (There should be benchmarks attached to the description)

I added a test (`TestMutexHasherConcurrentHash`) that verifies the thread safety of `MutexHasher`. If the mutex logic in `MutexHasher` were commented out or removed, the test would fail due to concurrency issues, and running the test with the `-race` flag would also detect data races.

Reply:
Followup question: are we actually using the `MutexHasher` anywhere (or planning to)?

Separate idea: I imagine it's probably slower, but have you tested a hash implementation that is inherently thread safe by always generating a new underlying hash object? For example:

```go
// Blaker2bHasher implements the Hasher interface using Blake2b algorithm.
type Blaker2bHasher struct{}

// NewBlaker2bHasher creates a new Blaker2bHasher.
func NewBlaker2bHasher() Blaker2bHasher {
	return Blaker2bHasher{}
}

func (Blaker2bHasher) Hash(data []byte) ([]byte, error) {
        h, _ := blake2b.New256(nil)
        _, _ = h.Write(data)
        return h.Sum(nil), nil
}
```

Reply:
Yea, I had started with that, but after I ran benchmarks comparing the two approaches:
The thread-safe version (creating a new hash object each time) is indeed about 5x slower and causes an allocation on every call, versus a single allocation for the reusable version.